arbiter/arbiter-core/src/main/java/org/deeplearning4j/arbiter/optimize/generator/GeneticSearchCandidateGenerator.java

```diff
@@ -0,0 +1,175 @@
+         */
+        public Builder(ParameterSpace<?> parameterSpace, ScoreFunction scoreFunction) {
+            this.parameterSpace = parameterSpace;
+            this.minimizeScore = scoreFunction.minimize();
+        }
+
+        /**
+         * @param populationModel The PopulationModel instance to use.
+         */
+        public Builder populationModel(PopulationModel populationModel) {
+            this.populationModel = populationModel;
+            return this;
+        }
+
+        /**
+         * @param selectionOperator The SelectionOperator to use. Default is GeneticSelectionOperator
+         */
+        public Builder selectionOperator(SelectionOperator selectionOperator) {
+            this.selectionOperator = selectionOperator;
+            return this;
+        }
+
+        public Builder dataParameters(Map<String, Object> dataParameters) {
+
+            this.dataParameters = dataParameters;
+            return this;
+        }
+
+        public GeneticSearchCandidateGenerator.Builder initDone(boolean initDone) {
+            this.initDone = initDone;
+            return this;
+        }
+
+        /**
+         * @param chromosomeFactory The ChromosomeFactory to use
+         */
+        public Builder chromosomeFactory(ChromosomeFactory chromosomeFactory) {
+            this.chromosomeFactory = chromosomeFactory;
+            return this;
+        }
+
+        public GeneticSearchCandidateGenerator build() {
+            if(populationModel == null) {
+                PopulationInitializer defaultPopulationInitializer = new EmptyPopulationInitializer();
+                populationModel = new PopulationModel.Builder(defaultPopulationInitializer).build();
+            }
+
+            if(chromosomeFactory == null) {
+                chromosomeFactory = new ChromosomeFactory();
+            }
+
+            if(selectionOperator == null) {
+                selectionOperator = new GeneticSelectionOperator.Builder().build();
+            }
+
+            return new GeneticSearchCandidateGenerator(this);
+        }
+    }
+
+    public final PopulationModel populationModel;
+    public final ChromosomeFactory chromosomeFactory;
+    public final SelectionOperator selectionOperator;
+
+    private GeneticSearchCandidateGenerator(Builder builder) {
+        super(builder.parameterSpace, builder.dataParameters, builder.initDone);
+
+        initialize();
+
+        chromosomeFactory = builder.chromosomeFactory;
+        populationModel = builder.populationModel;
+        selectionOperator = builder.selectionOperator;
+
+        chromosomeFactory.initializeInstance(builder.parameterSpace.numParameters());
+        populationModel.initializeInstance(builder.minimizeScore);
+        selectionOperator.initializeInstance(populationModel, chromosomeFactory);
+
+    }
+
+    @Override
+    public boolean hasMoreCandidates() {
+        return true;
```

Comment:
There's probably some edge cases where this isn't true.
Consider the case where there's 1 discrete parameter to optimize, with possible values {A,B,C}.
In this case, there's only 3 candidates - and the candidate generator should terminate once all 3 have been generated.